tests/gstreamer_mqtt/unittest_mqtt_w_helper.cc

```diff
@@ -0,0 +1,886 @@
+      &fdata);
+
+    return MQTTASYNC_FAILURE;
+  }
+
+  ret = std::async (std::launch::async, options->onSuccess, ctx,
+      &data);
+
+  return MQTTASYNC_SUCCESS;
+}
+
+/**
+ * @brief A mock function for MQTTAsync_destroy() in paho-mqtt-c
+ */
+void MQTTAsync_destroy (MQTTAsync *handle)
+{
+  return;
+}
+
+/**
+ * @brief A mock function for MQTTAsync_subscribe() in paho-mqtt-c
+ */
+int MQTTAsync_subscribe (MQTTAsync handle, const char * topic, int qos,
+    MQTTAsync_responseOptions * response)
+{
+  MQTTAsync_successData data;
+  std::future<void> ret;
+  void *ctx = response->context;
+
+  if (GstMqttTestHelper::getInstance ().getFailSubscribe ()) {
+    MQTTAsync_failureData fdata;
+
+    fdata.code = -1;
+    fdata.message = "";
+    ret = std::async (std::launch::async, response->onFailure, ctx, &fdata);
+    return MQTTASYNC_FAILURE;
+  }
+
+  ret = std::async (std::launch::async, response->onSuccess, ctx, &data);
+  return MQTTASYNC_SUCCESS;
+}
+
+/**
+ * @brief A mock function for MQTTAsync_unsubscribe() in paho-mqtt-c
+ */
+int MQTTAsync_unsubscribe (MQTTAsync handle, const char * topic,
+    MQTTAsync_responseOptions * response)
+{
+  void *ctx = response->context;
+  MQTTAsync_successData data;
+  std::future<void> ret;
+
+  if (GstMqttTestHelper::getInstance ().getFailUnsubscribe ()) {
+    MQTTAsync_failureData fdata;
+
+    fdata.code = -1;
+    fdata.message = "";
+    ret = std::async (std::launch::async, response->onFailure, ctx, &fdata);
+    return MQTTASYNC_FAILURE;
+  }
+
+  ret = std::async (std::launch::async, response->onSuccess, ctx, &data);
+  return MQTTASYNC_SUCCESS;
+}
+
+/**
+ * @brief A helper function to fill the timestamp information into the header
+ */
+void _set_ts_gst_mqtt_message_hdr (GstElement *elm, GstMQTTMessageHdr *hdr,
+    const GstClockTimeDiff diff_sent, const GstClockTime duration)
+{
+  GstClockTime base_time;
+  GstClockTime cur_time;
+  GstClockTimeDiff diff;
+  GstClock *clock;
+
+  hdr->base_time_epoch = GST_CLOCK_TIME_NONE;
+  clock = gst_test_clock_new ();
+  base_time = gst_element_get_base_time (elm) + diff_sent;
+  cur_time = gst_clock_get_time (clock);
+  gst_object_unref (clock);
+
+  diff = GST_CLOCK_DIFF (base_time, cur_time);
+  hdr->base_time_epoch = g_get_real_time () * GST_US_TO_NS_MULTIPLIER - diff;
+  hdr->sent_time_epoch = hdr->base_time_epoch + diff_sent;
+  hdr->duration = duration;
+}
+
+/**
+ * @brief Test for mqttsink with GstMqttTestHelper (push a GstBuffer)
+ */
+TEST (testMqttSink, sink_push0)
+{
+  GstHarness *h = gst_harness_new ("mqttsink");
+  GstFlowReturn ret;
+
+  g_object_set (h->element, "debug", true, NULL);
+  gst_harness_add_src_parse (h, "videotestsrc is-live=1 ! queue", TRUE);
+  GstMqttTestHelper::getInstance ().initFailFlags ();
+  ret = gst_harness_push_from_src (h);
+
+  EXPECT_EQ (ret, GST_FLOW_OK);
+
+  gst_harness_teardown (h);
+}
+
+/**
+ * @brief Test for mqttsink with GstMqttTestHelper (push a GstBuffer)
+ */
+TEST (testMqttSink, sink_push1)
```

Comment:
There seems to be no difference from `TEST (testMqttSink, sink_push0)`.
Please check it.

Reply:
This seems to be a mistake. Corrected. Thanks.